src/components/sourcepublisher/SourcePublisher.js

```diff
@@ -52,43 +28,29 @@ function reset() {
   PubSub.publish(RESET, {});
 }
 
-function publishLayer(data, type, name, url) {
-  const schema = typeToSchema[type];
-  if (!schema) {
-    throw Error(`No schema for ${type}`);
-  }
-  const validate = new Ajv().compile(schema);
-  const valid = validate(data);
-  if (!valid) {
-    const failureReason = JSON.stringify(validate.errors, null, 2);
-    warn(`Error while validating ${name}. Details in console.`);
-    console.warn(`"${name}" (${type}) from ${url}: validation failed`, failureReason);
-  }
-  PubSub.publish(typeToEvent[type], { data, url });
-  info(name);
-}
-
 function loadLayer(layer) {
   const {
-    name, type, url, requestInit = {},
+    name, type, url,
   } = layer;
-  fetch(url, requestInit)
-    .then((response) => {
-      if (response.ok) {
-        response.json().then((data) => {
-          publishLayer(data, type, name, url);
-        }, (failureReason) => {
-          warn(`Error while parsing ${name}.`);
-          console.warn(`"${name}" (${type}) from ${url}: parse failed`, failureReason);
-        });
-      } else {
-        warn(`Error HTTP status from ${name}.`);
-        console.warn(`"${name}" (${type}) from ${url}: HTTP failed`, response.headers);
-      }
-    }, (failureReason) => {
-      warn(`Error while fetching ${name}.`);
-      console.warn(`"${name}" (${type}) from ${url}: fetch failed`, failureReason);
-    });
+
+  // Iterate over loaders in order,
+  // and choose the best/first match based on the URL file extension.
+  const match = Array.from(extensionToLoader.entries())
+    .find(([extension]) => url.split('?')[0].endsWith(extension));
```

Comment:
I have made https://github.com/hubmapconsortium/vitessce/pull/691 to try to resolve this (that PR is a PR into this PR)

Reply:
Great, thanks! FWIW the `raster.json` has a `type` for each image type we support. Not sure if there is a way that we could make make a single manifest for expression data that ultimately points to these different sources, similar to `raster.json`.

Reply:
I had not thought of doing that, that is something to consider. I would prefer not to add the extra level of indirection that comes from the manifest-type files, but I am open to it if we decide that is the better way to go.

But I do not think this will solve the underlying problem, since creating a new manifest file would also require updating re-running a pipeline, and at that point, I would just update the pipeline to produce the zarr files

Reply:
I agree the extra level of indirection is not really favorable, but my thought currently for the image data (at least) is that `raster.json` contains metadata and source urls to create in-memory javascript objects that have a shared interface. 

It might be worth describing this more formally (to help me think about whats going on) and whether we can provide something similar for other datatypes:

Top-level `raster.json`: manifest that contains a list of image `sources` (`url`, `metadata`, `imageType`).

[`initLoader`](https://github.com/hubmapconsortium/vitessce/blob/e51894ef90e24a83a06fa3dd06354635792aa272/src/components/layer-controller/LayerControllerSubscriber.js#L31): Takes an image source (as described in the manifest) and returns an `loader` object which has a shared interface for both OME-TIFF and Zarr.

Reply:
I agree, it doesn't solve the underlying problem, but by publishing an object that describes the heatmap data (rather than loading all the data in the `sourcePublisher`) to the `HeatmapSubscriber`, we can isolate creating loaders for the different heatmap data formats in `Subscriber`.

Reply:
I understand that for images, `metadata` is an important requirement which necessitates the `raster.json` indirection, since this metadata does not (cannot?) exist in certain image file types. But for these files for which I have added the loaders, the only information required to choose a loader is a `fileType` field, since we have control over the internals of these files (with zarr and json schemas) to add arbitrary metadata inside.